### ISSUE TITLE
False warning HAA0101 when params follows an optional parameter

### ISSUE THREAD

**Issue by KalleOlaviNiemitalo** [microsoft/RoslynClrHeapAllocationAnalyzer#84]
When a function has an optional parameter followed by a `params` parameter, and it is called without the corresponding arguments, the analyzer reports warning HAA0101 (array allocation for `params`) even though the generated code uses `Array.Empty<T>` instead of allocating an array.

Related to <https://github.com/microsoft/RoslynClrHeapAllocationAnalyzer/pull/78>.

# To reproduce

DefaultParamsTest.csproj:

```XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Library</OutputType>
    <TargetFramework>netstandard2.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="ClrHeapAllocationAnalyzer" Version="3.0.0" />
  </ItemGroup>

</Project>
```

Demo.cs:

```C#
namespace DefaultParamsTest
{
    static class Demo
    {
        static void Fun1()
        {
            Fun2();
        }

        static void Fun2(int i = 0, params object[] args)
        {
        }
    }
}
```

Build with .NET SDK 3.1.301:

```
dotnet build -p:Configuration=Release
```

# Expected behavior

No warnings.

# Actual behavior

HAA0101 warning:

```
Microsoft (R) Build Engine version 16.6.0+5ff7b0c9e for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
Demo.cs(7,13): warning HAA0101: This call site is calling into a function with a 'params' parameter. This results in an array allocation […\DefaultParamsTest.csproj]
  DefaultParamsTest -> …\bin\Release\netstandard2.1\DefaultParamsTest.dll

Build succeeded.

Demo.cs(7,13): warning HAA0101: This call site is calling into a function with a 'params' parameter. This results in an array allocation […\DefaultParamsTest.csproj]
    1 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.09
```

IL disassembly shows no array allocation:

```
.method private hidebysig static void  Fun1() cil managed
{
  // Code size       12 (0xc)
  .maxstack  8
  IL_0000:  ldc.i4.0
  IL_0001:  call       !!0[] [netstandard]System.Array::Empty<object>()
  IL_0006:  call       void DefaultParamsTest.Demo::Fun2(int32,
                                                         object[])
  IL_000b:  ret
} // end of method Demo::Fun1
```

**Comment by Youssef1313** [comment]
Note: the fix is **not** to check for `arguments.Count <= methodInfo.Parameters.Length - 1`.

Consider:

```
namespace DefaultParamsTest
{
    static class Demo
    {
        static void Fun1()
        {
            Fun2(args: "");
        }

        static void Fun2(int i = 0, params object[] args)
        {
        }
    }
}
```

**Comment by Youssef1313** [comment]
Note 2:

Before Array.Empty<T> existed, which was around .NET Framework 4.5, an empty array was allocated. So, this **must** check whether `Array.Empty<T>` exists in the compilation or not.
